### PR TITLE
Speed up snapshot shards service by avoiding looping all shards needlessly

### PR DESCRIPTION
**Pull request by original-brownbear**
Store a flag to indicate whether or not we have INIT stage shards in a snapshot to avoid looping over those snapshots that don't needlessly in the snapshot shards service.
We loop all shards when constructing the snapshot entry anyway so this is almost free when constructing the entry, but saves considerable CPU time when there's many queued up snapshots during CS application.

relates #77466

### CODE REVIEW

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/part-3 (unrelated and known)

File: server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java [hunk 1]
```diff
@@ -704,13 +711,16 @@ public static Entry snapshot(
...
+            boolean hasInitStateShards = false;
             for (Map.Entry<ShardId, ShardSnapshotStatus> entry : shards.entrySet()) {
                 final ShardId shardId = entry.getKey();
                 final IndexId indexId = indices.get(shardId.getIndexName());
                 final Index index = shardId.getIndex();
                 final Index existing = res.put(indexId.getName(), index);
                 assert existing == null || existing.equals(index) : "Conflicting indices [" + existing + "] and [" + index + "]";
-                byRepoShardIdBuilder.put(new RepositoryShardId(indexId, shardId.id()), entry.getValue());
+                final var shardSnapshotStatus = entry.getValue();
+                hasInitStateShards |= shardSnapshotStatus.state() == ShardState.INIT;
```
Comment (pxsalehi) [inline server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java:722]: Nit: call me old-fashioned, but I find an if-statement more readable. :)

File: server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java [hunk 2]
```diff
@@ -686,6 +686,13 @@ public static class Entry implements Writeable, ToXContent, RepositoryOperation,
         @Nullable
         private final String failure;
 
+        /**
+         * Flag set to true in case any of the shard snapshots in {@link #shards} are in state {@link ShardState#INIT}.
+         * This is used by data nodes to determine if there is any work to be done on a snapshot by them without having to iterate
+         * the full {@link #shards} map.
+         */
+        private final boolean hasInitStateShards;
```
Comment (pxsalehi) [inline server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java:694]: Nit: I find `hasShardsInInitState` easier to understand.

**Review by pxsalehi — APPROVED** [review]
LGTM! Seems like a reasonable optimization. But feel free to wait for another LGTM, as I might have missed something, since the (rest of the) code is a bit complicated.

**Comment by pxsalehi** [comment]
Also one question. Would adding/updating tests in `SnapshotsInProgressSerializationTests` be of any value here? Or is that too trivial?

**Comment by original-brownbear** [comment]
Thanks Pooya!

**Pull request merged**